<a href="https://colab.research.google.com/github/rodrigoblopes/landscape/blob/master/My_QuickNotes_Panda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intro to Pandas


In [1]:
#Import pandas library
import pandas as pd

#Download Data Sample
From [Yahoo Finance](https://finance.yahoo.com/quote/BHP.AX/history?p=BHP.AX), download the Historical Data of a Stock option. In this example, we are using the BHP.AX.

Once you download, upload the file on the **Files** container on the left side.
Next step, click over the file and choose **Copy path**

In [2]:
#import csv file to Pandas
#create data frame (df) variable
df = pd.read_csv("/content/BHP.AX.csv")

In [3]:
#show the dataframe like a spreadsheet
df.head()

Date       Open       High        Low      Close  Adj Close    Volume
0  2022-01-17  46.790001  47.000000  45.930000  46.150002  41.378586  12201119
1  2022-01-18  46.439999  46.990002  46.189999  46.700001  41.871719   9940912
2  2022-01-19  46.400002  46.880001  46.150002  46.560001  41.746193  12821922
3  2022-01-20  47.209999  48.240002  47.209999  48.009998  43.046280  12839430
4  2022-01-21  47.000000  47.000000  45.570000  45.700001  40.975109  20068064

In [4]:
#check the variable type df
type(df)

pandas.core.frame.DataFrame

In [5]:
#check the series type inside df variable
type(df["Open"])

pandas.core.series.Series

In [6]:
#see the column type of df variable
df["Open"]

0      46.790001
1      46.439999
2      46.400002
3      47.209999
4      47.000000
         ...    
247    47.580002
248    48.650002
249    49.000000
250    49.250000
251    49.810001
Name: Open, Length: 252, dtype: float64

In [7]:
#check the dataframe size (format)
#first number: lines
#second number: series (columns)
df.shape

(252, 7)

#Knowing the data

The file above it has 252 lines and 7 columns. 

When we import a *dataset* we are looking for a quick analysis in some entries, so we can have a roughly idea about it.

* `df.head()` - show first 5 entries from the dataframe
* `df.tail()` - show last 5 entries from the dataframe

In [8]:
#show the first 10 entries
df.head(10)

Date       Open       High        Low      Close  Adj Close  \
0  2022-01-17  46.790001  47.000000  45.930000  46.150002  41.378586   
1  2022-01-18  46.439999  46.990002  46.189999  46.700001  41.871719   
2  2022-01-19  46.400002  46.880001  46.150002  46.560001  41.746193   
3  2022-01-20  47.209999  48.240002  47.209999  48.009998  43.046280   
4  2022-01-21  47.000000  47.000000  45.570000  45.700001  40.975109   
5  2022-01-24  45.320000  45.680000  44.900002  45.599998  40.885448   
6  2022-01-25  45.240002  45.650002  44.910000  45.029999  40.374378   
7  2022-01-27  45.299999  46.485001  45.049999  45.669998  40.948208   
8  2022-01-28  47.259998  47.259998  46.099998  46.919998  42.068970   
9  2022-01-31  46.049999  46.360001  45.549999  46.349998  41.557903   

      Volume  
0   12201119  
1    9940912  
2   12821922  
3   12839430  
4   20068064  
5   24745712  
6   23844958  
7   27712917  
8  327417026  
9   37660130

In [9]:
#show the last 10 entries
df.tail(10)

Date       Open       High        Low      Close  Adj Close  \
242  2023-01-03  46.099998  46.099998  44.930000  45.330002  45.330002   
243  2023-01-04  45.299999  46.180000  45.250000  46.040001  46.040001   
244  2023-01-05  45.980000  46.240002  45.680000  46.049999  46.049999   
245  2023-01-06  47.110001  47.779999  46.959999  47.509998  47.509998   
246  2023-01-09  48.000000  48.430000  47.880001  48.000000  48.000000   
247  2023-01-10  47.580002  48.049999  47.520000  47.799999  47.799999   
248  2023-01-11  48.650002  48.930000  48.490002  48.509998  48.509998   
249  2023-01-12  49.000000  49.500000  48.720001  49.380001  49.380001   
250  2023-01-13  49.250000  49.939999  49.250000  49.639999  49.639999   
251  2023-01-16  49.810001  50.000000  49.360001  49.689999  49.689999   

      Volume  
242  4480991  
243  7802250  
244  7536105  
245  8106986  
246  5757435  
247  5653577  
248  7432150  
249  8825502  
250  8519896  
251  7439473

In [10]:
#check the columns names
#create variable column_names
column_names = df.columns

#print column_names
column_names

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume'], dtype='object')

Know the type of data is essential.


*   Financial: ```float``` (number after "," or "." - depending the currency)
*   Year: ```int```
*   Full date: ```datetime```

This is important so we can manipulate the *dataset*

To know all types of variable for each columns (series), use ```df.types```




In [12]:
df.dtypes

Date          object
Open         float64
High         float64
Low          float64
Close        float64
Adj Close    float64
Volume         int64
dtype: object

Another way to figure out the type of variable and check the null values is using ```df.info()```

In [ ]:
df.info()

# Choosing a column (series)
There are different ways to choose a series inside the Dataframe.

In the majority of cases, we want to choose only one column.

To choose the column, you must insert between square brackets [ ]

In [15]:
#select the series "High" from our dataframe
df["High"]

0      47.000000
1      46.990002
2      46.880001
3      48.240002
4      47.000000
         ...    
247    48.049999
248    48.930000
249    49.500000
250    49.939999
251    50.000000
Name: High, Length: 252, dtype: float64

If the column name has no spaces or special character, you can use the following:
Instead ```df.["High"]``` you can use ```df.High```. The result will be the same.

In [17]:
#select the series "High"from our dataframe
df.High

0      47.000000
1      46.990002
2      46.880001
3      48.240002
4      47.000000
         ...    
247    48.049999
248    48.930000
249    49.500000
250    49.939999
251    50.000000
Name: High, Length: 252, dtype: float64

# Calc the column average
When we run the code ```df.High```, you'll see the column printed from ```High``` values. To find the average, you need to use ```df.High.mean()```


In [18]:
# Calc the average for the column "High"
df.High.mean()

44.07590959126984

#Working with dates (```datetime```)
When you import a file with dates, probably Pandas won't recognize the format automatically.

When you checked ```df.info()```, you saw the column ```Date``` was as ```object```. In this case, the columns it's text (```string```) and you can't extract useful stuff.

Let's convert the column using ```pd.to_datetime``` and set the date format.

In [19]:
#before
df.Date

0      2022-01-17
1      2022-01-18
2      2022-01-19
3      2022-01-20
4      2022-01-21
          ...    
247    2023-01-10
248    2023-01-11
249    2023-01-12
250    2023-01-13
251    2023-01-16
Name: Date, Length: 252, dtype: object

In [ ]:
#convert column Date in datetime
df.Date = pd.to_datetime(df.Date, format="%Y-%m-%d")

In [ ]:
# after
df.Date

Now it's possible to access on specific components from the date, using ```df.Date.dt```

In [23]:
df.Date.dt.year

0      2022
1      2022
2      2022
3      2022
4      2022
       ... 
247    2023
248    2023
249    2023
250    2023
251    2023
Name: Date, Length: 252, dtype: int64

In [24]:
df.Date.dt.month

0      1
1      1
2      1
3      1
4      1
      ..
247    1
248    1
249    1
250    1
251    1
Name: Date, Length: 252, dtype: int64

In [27]:
df.Date.dt.day

0      17
1      18
2      19
3      20
4      21
       ..
247    10
248    11
249    12
250    13
251    16
Name: Date, Length: 252, dtype: int64

There are different formats an user can insert. You can find for instance:

*   31/06/15
*   31/06/2015
*   31-6-15
*   2015-6-31
*   31 June 2015

To deal with this, you need to understand the documentation. Useful links:

* https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html?highlight=to_datetime#pandas.to_datetime
* https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior

